src/systems/environmental_sensor_system/EnvironmentalSensorSystem.cc

```diff
@@ -113,30 +185,72 @@ class EnvironmentalSensor : public gz::sensors::Sensor
   {
     if (!this->ready) return false;
 
-    if (!this->session.has_value()) return false;
 
-    // Step time if its not static
-    if (!this->gridField->staticTime)
-      this->session = this->gridField->frame[this->fieldName].StepTo(
-        this->session.value(), std::chrono::duration<double>(_now).count());
+    std::optional<double> dataPoints[3];
+    for (std::size_t i = 0; i < this->numberOfFields; ++i)
+    {
+      if (this->fieldName[i] == "")
+      {
+        // Empty field name means the column should default to zero.
+        dataPoints[i] = 0;
+        continue;
+      }
+
+      if (!this->session[i].has_value()) return false;
+
+      // Step time if its not static
+      if (!this->gridField->staticTime)
+        this->session[i] = this->gridField->frame[this->fieldName[i]].StepTo(
+          this->session[i].value(),
+          std::chrono::duration<double>(_now).count());
 
-    if (!this->session.has_value()) return false;
+      if (!this->session[i].has_value()) return false;
 
-    gz::msgs::Double msg;
-    *msg.mutable_header()->mutable_stamp() = gz::msgs::Convert(_now);
-    auto frame = msg.mutable_header()->add_data();
-    frame->set_key("frame_id");
-    frame->add_value((this->frameId == "") ? this->Name() : this->frameId);
-    auto data = this->gridField->frame[this->fieldName].LookUp(
-      this->session.value(), this->position);
-    if (!data.has_value())
+      dataPoints[i] = this->gridField->frame[this->fieldName[i]].LookUp(
+        this->session[i].value(), this->position);
+    }
+
+    if(this->numberOfFields == 1) {
```

Comment:
@arjo129 nit: 
```suggestion
    if (this->numberOfFields == 1) {
```
Have you ran `codecheck`?